src/hotspot/share/code/nmethod.cpp

```diff
@@ -1262,20 +1316,21 @@ nmethod::nmethod(
     _deopt_mh_handler_offset = 0;
     _unwind_handler_offset   = 0;
 
-    _metadata_offset         = checked_cast<uint16_t>(align_up(code_buffer->total_oop_size(), oopSize));
-    _dependencies_offset     = checked_cast<uint16_t>(_metadata_offset + align_up(code_buffer->total_metadata_size(), wordSize));
-    _scopes_pcs_offset       = _dependencies_offset;
+    CHECKED_CAST(_metadata_offset, uint16_t, (align_up(code_buffer->total_oop_size(), oopSize)));
+    DEBUG_ONLY( int data_end_offset = _metadata_offset + align_up(code_buffer->total_metadata_size(), wordSize); )
+    assert((data_offset() + data_end_offset) <= nmethod_size, "wrong nmethod's size: %d < %d", nmethod_size, (data_offset() + data_end_offset));
+
+    // native wrapper does not have read-only data but we need unique not null address
+    _immutable_data          = data_end();
```

Comment:
I can't use nullptr because VM expects not null address when it checks, for example, `dependencies_begin()` even so sizes are 0.
I used `data_end()` instead of nullptr in other places too.